### PR TITLE
Recording Plane

### PR DESCRIPTION
**Pull request by RevathiJambunathan**
## Summary

There are two major parts in this PR. The first is a new diagnostic type, RecordingPlane, and the second is to use the recording plane data in a boosted frame simulation.

 Here is an example of using the new diagnostic.
  ```
  rp.diag_type = RecordingPlane
  rp.file_prefix = ./extractref/station
  rp.station_location = 1.e-4
  rp.species = beam
  ```
With this new diagnostic, we can record a time history of the fields and particles at a specified location. The field data are interpolated to nodal positions. This makes it easier to do Lorentz transformation on the data.
 
For a restart simulation in a boosted frame, runtime parameter `warpx.impose_field_file_path` can be used to specify the path to the field data and we use "laser" particles to generate the fields.
  ```
  warpx.impose_field_file_path = ".extractref/station"
  lasers.names = laser_y
  laser_y.position     = 0. 0. 1.e-4  # must be consist with the lab frame run
  laser_y.direction    = 0. 0. 1.
  laser_y.polarization = 0. 1. 0.
  laser_y.e_max        = 1.e12
  laser_y.wavelength   = 1.0e-6
  laser_y.profile      = station_field
  laser_y.station_file_name = "./extractref/station"
  ```

As for beam particles in the restart run, we can use the existing capability of injection from an external file and set
  ```
  particles.rigid_injected_species = beam
  beam.injection_style = external_file
  beam.injection_file = "./extractref/station/openpmd_000000.bp"
  beam.injection_from_recording_plane = true
  beam.zinject_plane = 1.e-4  # must be consist with the lab frame run
  beam.rigid_advance = true
  ```

Note that the z-direction values of `geometry.prob_lo` and `prob_hi` in the inputs file are ignored for the restart run, and they are set automatically using the recording plane data.

## Known limitations.
  * 1D not supported
  * Single level only
  * Single recording plane only

## TO-DO in the future
  * Get `beam.zinjection_plane` from the data
  * Get `laser_y.position = "0. 0. 1.e-4"` from the data
  * Set laser's `e_max` using the field data
  * Remove duplication in setting `laser.get station_file_name` and `warpx.impose_field_file_path`
  * Write field data in openPMD format

### CODE REVIEW

**Comment by RevathiJambunathan** [comment]
Meeting notes : 
Store time at which particles cross the plane - as run-time attribute
Similar to scraping diags : use `WriteToFile`

**Comment by WeiqunZhang** [comment]
Mark it "Ready for review" to trigger CIs.

File: Source/Diagnostics/MultiDiagnostics.cpp [hunk 1]
```diff
@@ -82,11 +86,11 @@ MultiDiagnostics::FilterComputePackFlush (int step, bool force_flush, bool BackT
     int i = 0;
     for (auto& diag : alldiags){
         if (BackTransform) {
-            if (diags_types[i] == DiagTypes::BackTransformed) {
+            if ( (diags_types[i] == DiagTypes::BackTransformed) || (diags_types[i] == DiagTypes::RecordingPlane)) {
```
Comment (WeiqunZhang) [inline Source/Diagnostics/MultiDiagnostics.cpp:89]: The logic here is hard to follow. At least the name `BackTransform` is misleading.

File: Source/Diagnostics/RecordingPlaneDiagnostics.H [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+#endif
+    StationNormalDir m_station_normal = StationNormalDir::z;
+
+    /** whether to flush at this time step
+     * \param[in] step current time step
+     * \param[in] i_buffer index of a back-transformed snapshot
+     * \param[in] force_flush if true, return true for any step
+     * \return bool, whether to flush
+     */
+    bool DoDump (int step, int i_buffer, bool force_flush=false) override;
```
Comment (WeiqunZhang) [inline Source/Diagnostics/RecordingPlaneDiagnostics.H:42]: A comment not specific to this PR. A lot of virtual functions in Diagnostics have default parameters. This is probably a bad idea. We might consider change it. @ax3l Maybe we should discuss it in the next meeting.

 https://google.github.io/styleguide/cppguide.html#Default_Arguments

**Review by WeiqunZhang — DISMISSED** [review]

**Comment by ax3l** [comment]
Reproducer for MPI bug by @jlvay : `If running with 1 MPI processes, I see the output of 10 particles, as expected. If running with 2 MPI processes, I see only 2.`
```
stop_time=4.16717648325019e-13

# algo
algo.maxwell_solver = CKC
algo.particle_pusher = vay
algo.particle_shape = 3

# amr
#amr.blocking_factor = 2
#amr.max_grid_size = 64
#warpx.numprocs = 1 2
amr.max_level = 0
amr.n_cell = 64 64

# beam
beam.charge = -q_e
beam.initialize_self_fields = 0
beam.injection_style = gaussian_beam
beam.mass = m_e
# beam.momentum_distribution_type = constant
# beam.ux = 0.0
# beam.uy = 0.0
# beam.uz = 10.954451150103322
beam.momentum_distribution_type = gaussian
beam.ux_m = 0.0
beam.uy_m = 0.0
beam.uz_m = 3.
beam.ux_th = 0.2
beam.uy_th = 0.2
beam.uz_th = 1.
beam.npart = 10
beam.q_tot = -10e-15
beam.rigid_advance = 1
beam.x_m = 0.0
beam.x_rms = 1e-06
beam.y_m = 0.0
beam.y_rms = 1e-06
beam.z_m = -1e-06
beam.z_rms = 1.e-07
beam.zinject_plane = 0.0

# boundary
boundary.field_hi = pml pml
boundary.field_lo = pml pml
boundary.particle_hi = absorbing absorbing
boundary.particle_lo = absorbing absorbing

# diagnostics
diagnostics.diags_names = diag1 diag2 rp

diag1.diag_type = Full
diag1.file_prefix = ./extractref/ref_diags_beam
diag1.format = openpmd
diag1.openpmd_backend = bp
diag1.intervals = 152
diag1.species = beam
diag1.fields_to_plot = none

diag2.fields_to_plot = Ex Ey Ez Bx By Bz jx jy jz rho rho_beam
diag2.diag_type = Full
diag2.file_prefix = ./extractref/ref_diags_field
diag2.format = openpmd
diag2.openpmd_backend = bp
#diag2.plot_raw_fields = true
diag2.intervals = 152
diag2.write_species = 0

# geometry
geometry.dims = 2
geometry.prob_hi = 1.825741858350554e-05 1.8257418583505539e-07
geometry.prob_lo = -1.825741858350554e-05 -1.0954451150103323e-05

# my_constants
my_constants.Rc = 6.390096504226938e-06
my_constants.kp = 1030698.0524126256
my_constants.n0 = 3e+25

# particles
particles.rigid_injected_species = beam
particles.species_names = beam
particles.use_fdtd_nci_corr = 0

# warpx
warpx.boost_direction = z
warpx.cfl = 1.0
warpx.compute_max_step_from_btd = 0
warpx.do_dive_cleaning = 0
warpx.do_moving_window = 1
warpx.filter_npass_each_dir = 1 1
warpx.gamma_boost = 1.0
warpx.moving_window_dir = z
warpx.moving_window_v = 1.0
warpx.serialize_initial_conditions = 1
warpx.use_filter = 1
warpx.use_filter_compensation = 0
warpx.verbose = 1

rp.diag_type = RecordingPlane
rp.file_prefix = ./extractref/station
rp.station_location = 0.5e-4
rp.intervals = 0
rp.buffer_size = 1024000
rp.format = openpmd
rp.species = beam
rp.openpmd_backend = bp

warpx.deposit_before_plane = 0
```

Visible via:
```py
from openpmd_viewer import OpenPMDTimeSeries

ts = OpenPMDTimeSeries('./extractref/station/')
t,x,z,ux,uy,uz = ts.get_particle( ['time','x','z','ux','uy','uz'], species = 'beam', iteration = 0)
print (x)
```

**Comment by RevathiJambunathan** [comment]
[7e8d149](https://github.com/ECP-WarpX/WarpX/pull/4119/commits/7e8d14967fc346694eaf6330286152fe6572c2d1) 
this commit fixes the error
now we see 10 particles with 1 or more mpi procs
